app/pipeline_factory.cpp

```diff
@@ -8,57 +8,140 @@
 #include "descripting/cog.h"
 #include "matching/nearest_neighbour.h"
 #include "matlab_reader.h"
+#include "matlab_reader_concurrent.h"
 #include "point_cloud_filtering/subsample.h"
 #include "registration/disparity_registration.h"
+#include "registration/disparity_registration_cpu_optimized.h"
 #include "trajectory_builder/cog_trajectory_builder.h"
 #include <boost/log/trivial.hpp>
 
 namespace MouseTrack {
 
 Pipeline
 PipelineFactory::fromCliOptions(const op::variables_map &options) const {
-  std::string src_dir;
-  if (options.count("src-dir") > 0) {
-    src_dir = options["src-dir"].as<std::string>();
-  } else {
-    src_dir = "";
+  if (options.count("src-dir") == 0) {
     BOOST_LOG_TRIVIAL(info) << "No command line source path given.";
   }
-  std::unique_ptr<MatlabReader> reader{new MatlabReader(src_dir)};
-  if (options.count("first-frame")) {
-    reader->setBeginFrame(
-        std::max(reader->beginFrame(), options["first-frame"].as<int>()));
-  }
-  if (options.count("last-frame")) {
-    reader->setEndFrame(
-        std::min(reader->endFrame(), options["last-frame"].as<int>() + 1));
-  }
-  std::unique_ptr<Registration> registration{new DisparityRegistration()};
+  std::unique_ptr<Reader> reader = getReader(options);
+
+  std::unique_ptr<Registration> registration{getRegistration(options)};
+
   std::unique_ptr<PointCloudFiltering> cloudFiltering{
-      new SubSample(options["subsample-to"].as<int>())};
+      getCloudFiltering(options)};
 
   std::unique_ptr<Clustering> clustering{getClustering(options)};
 
-  std::unique_ptr<Descripting> descripting{new CenterOfGravity()};
-  std::unique_ptr<Matching> matching{new NearestNeighbour()};
+  std::unique_ptr<Descripting> descripting{getDescripting(options)};
+
+  std::unique_ptr<Matching> matching{getMatching(options)};
+
   std::unique_ptr<TrajectoryBuilder> trajectoryBuilder{
-      new CogTrajectoryBuilder()};
+      getTrajectoryBuilder(options)};
+
   return Pipeline(std::move(reader), std::move(registration),
                   std::move(cloudFiltering), std::move(clustering),
                   std::move(descripting), std::move(matching),
                   std::move(trajectoryBuilder));
 }
 
+std::unique_ptr<Reader>
+PipelineFactory::getReader(const op::variables_map &options) const {
```

Comment:
A little concerned with the lack of comments. It could be that this is really standard and obvious stuff, so if you feel there is no need for comments, then no worries.

Reply:
I'll look over it